### PR TITLE
wrote ssh exception handler

### PR DESCRIPTION
**Pull request by Hidanio**

### CODE REVIEW

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt [hunk 1]
```diff
@@ -144,10 +145,12 @@ class SshConnectionService(private var myProject: Project) : Disposable {
...
-                        mySftpChannel = myConnectionBuilder!!.openSftpChannel(2)
-                        mySftpClient = MySshUtils.getSftpClient(mySftpChannel!!)
+                        SshHandler.handle {
+                            mySftpChannel = myConnectionBuilder!!.openSftpChannel(2)
+                            mySftpClient = MySshUtils.getSftpClient(mySftpChannel!!)
 
-                        onSuccessful?.run()
+                            onSuccessful?.run()
+                        }
                     } catch (e: SshException) {
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:154]: You have unwrapped the exception in `SshHandler.unwrap`. Now the exception data will not be caught

File: src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        return try {
+            call()
+        } catch (e: SshException) {
+          throw  unwrap(e)
+        }
+    }
+
+    fun unwrap(ex: Exception): Throwable {
+        var cause = ex.cause
+        while(cause?.cause != null) {
```
[the remark this answers is not in the document]
Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt]: fixed

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt (hunk 1 above)
Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:154]: Fixed. Maybe later we’ll do a more detailed processing of each unwound exception if necessary, but for now it’s ok i think

File: src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        }
+    }
+
+    fun unwrap(ex: Exception): Throwable {
+        var cause = ex.cause
+        while (cause?.cause != null) {
+            cause = cause.cause
+        }
+
+        if (cause is TimeoutException) {
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt:21]: Let's not do it this way. It is better to move the check to `catch` where you call `SshHandler.handle`

File: src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        }
+    }
+
+    fun unwrap(ex: Exception): Throwable {
+        var cause = ex.cause
+        while (cause?.cause != null) {
+            cause = cause.cause
+        }
+
+        return cause?: ex
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt:20]: whitespace :(

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt [hunk 5]
```diff
@@ -185,7 +177,24 @@ class SshConnectionService(private var myProject: Project) : Disposable {
...
 
-                        LOG.warn("Failed to connect", e)
+                        LOG.warn("Failed to connect", ex)
+                    } catch (ex: TimeoutException) {
+                        if (indicator.isCanceled) {
+                            LOG.info("Cancelled by user", ex)
+                            return
+                        }
+
+                        AdmNotification("Don't forget to touch the yubikey if it blinks when using the AdmStorm plugin's features")
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:187]: Let's simplify the text

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt [hunk 6]
```diff
@@ -185,7 +177,24 @@ class SshConnectionService(private var myProject: Project) : Disposable {
...
+                            .withTitle("Yubikey waiting timeout")
+                            .withActions(AdmNotification.Action("Reconnect...") { _, notification ->
+                                notification.expire()
+                                connectWithConnector(connector, onSuccessful)
+                            }).show()
+
+                        LOG.info("Yubikey waiting timeout", ex)
+                    } catch (ex: Exception) {
+                        LOG.error("Unhandled exception ${ex.javaClass.name}")
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:196]: This is good, but in a good way it is necessary to inform the user that something went wrong.
I would just throw the exception further

File: src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt (hunk 4 above)
Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshHandler.kt:20]: Fixed

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt (hunk 5 above)
Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:187]: Simplified

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt (hunk 6 above)
Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:196]: Added notification

File: src/main/kotlin/com/vk/admstorm/utils/MySshUtils.kt [hunk 7]
```diff
@@ -109,17 +112,11 @@ object MySshUtils {
...
-            ?.replaceFirstChar { it.uppercaseChar() }
-            ?.plus("<br>")
-            ?: ""
-
-        val message = """
-            ${exceptionMessage}Plugin can try to automatically reset the Yubikey and reconnect or you can do it 
-            yourself with ${code("ssh-agent")} and push 'Reconnect' button.
-            """.trimIndent()
+    private fun handleSshException(project: Project, ex: Exception) {
+        val message =
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/utils/MySshUtils.kt:116]: Let's format it like this. It looks better
```kotlin
val message = "${ex.message}<br>" +
        "Plugin can try to automatically reset the Yubikey or you can do it yourself with " +
        code("ssh-agent")
```

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt [hunk 8]
```diff
@@ -185,7 +178,52 @@ class SshConnectionService(private var myProject: Project) : Disposable {
...
+                        AdmErrorNotification(message, true)
+                            .withTitle("Did you connect to corporate access system?")
+                            .withActions(AdmNotification.Action("Reconnect...") { _, notification ->
+                                notification.expire()
+                                connectWithConnector(connector, onSuccessful)
+                            }).show()
+                        LOG.info("Corporate access error", ex)
+                    } catch (ex: Exception) {
+                        val errorMessage = ex.javaClass.name
+                        LOG.error("Unhandled exception $errorMessage")
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:224]: Why not just `${ex.message}`?

Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt]: ok, expanded the logs a bit

File: src/main/kotlin/com/vk/admstorm/utils/MySshUtils.kt [hunk 9]
```diff
@@ -85,21 +86,23 @@ object MySshUtils {
...
             null
-        } catch (e: IllegalStateException) {
-            handleSshException(project, e)
+        } catch (ex: IllegalStateException) {
+            handleSshException(project, ex)
             null
-        } catch (e: Exception) {
-            handleSshException(project, e)
+        } catch (ex: Exception) {
+            LOG.error("Unhandled exception ${ex.javaClass.name}")
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/utils/MySshUtils.kt:105]: And then notify the user?

File: src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt [hunk 10]
```diff
@@ -185,7 +178,52 @@ class SshConnectionService(private var myProject: Project) : Disposable {
...
+
+                        AdmErrorNotification(message, true)
+                            .withTitle("Did you connect to corporate access system?")
+                            .withActions(AdmNotification.Action("Reconnect...") { _, notification ->
+                                notification.expire()
+                                connectWithConnector(connector, onSuccessful)
+                            }).show()
+                        LOG.info("Corporate access error", ex)
+                    } catch (ex: Exception) {
+                        val errorMessage = ex.javaClass.name
```
Comment (Danil42Russia) [inline src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt:223]: What for? Why not just `mesage`?

Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/ssh/SshConnectionService.kt]: Ok

File: src/main/kotlin/com/vk/admstorm/utils/MySshUtils.kt (hunk 9 above)
Comment (Hidanio) [reply src/main/kotlin/com/vk/admstorm/utils/MySshUtils.kt]: ok

**Review by Danil42Russia — APPROVED** [review]

**Pull request merged**